### ISSUE TITLE
Prevent duplicate fetch requests

### ISSUE THREAD

**Issue by andyrichardson** [andyrichardson/tipple#20]
useQuery requests should be deduped. 

This needs to be implemented in the executeRequest function using the URL as a key and returning existing promises that are in flight.

Any beginners who aren't sure where to start, leave a comment and we can run through this together or pair.

**Comment by simonxabris** [comment]
Hey

I'm be interested in solving this, this would be my first open source contribution so bare with me please 😅. 

I have a couple questions, by useQuery you mean the useFetch hook, right?
Also could you explain you solution idea a little further? I've looked at the source, I'm pretty sure I'm missing something obvious, but as I'v seen executeRequest is just a simple helper function that executes a fetch request and returns the json the request resolves with and as such I don't see how it has any knowledge of what requests have been sent earlier and what promises are pending.

Thanks for any advice and help!

**Comment by andyrichardson** [comment]
@simonxabris welcome aboard!

> by useQuery you mean the useFetch hook, right?

Yeah, my mistake (GraphQL uses the terms 'query' and 'mutation' to mean 'fetch' and 'push').

> I've looked at the source, I'm pretty sure I'm missing something obvious, but as I'v seen executeRequest is just a simple helper function that executes a fetch request and returns the json the request resolves with and as such I don't see how it has any knowledge of what requests have been sent earlier and what promises are pending.

You're spot on. There needs to be a way for 'executeRequest' to track in flight operations. This would be easier using streams but can be done with promises.

```
// Our collection to track in flight operations
const inflight = {};

const executeRequest = (...) => {
  // Check if current URL is in flight
  
  // If so, use that

  // If not, create a new request and add it to the inFlight collection

  // Wait for that request to complete and then remove from the inFlight collection

  // continue as normal 
}
```

I've tagged this as beginner but it's definitely on the trickier side. You're going the right way about this 👍

**Comment by simonxabris** [comment]
Thanks for the clarifications!

I'll try to find some time today to figure out a solution. I'll get back to you if I run into any trouble.

**Comment by simonxabris** [comment]
Soo I think I'm on the right path here.

What I have trouble with is I have this structure right now:
```javascript 
if (infLight[url]) {
// handle request which are in flight
} else {
// handle requests normally
}
``` 
and in the block to handle the requests which have already been made I'm a bit stuck because I don't know what I should be returning with.

I can't do

```javascript
const response = await inFlight[url];
const json = response.json();
return json
```
because by that time this Response have already been used by the initial request and I get 'Failed to execute 'json' on 'Response': body stream is locked' error, I also can't do

```javascript
const response = await inFlight[url];
const responseClone = response.clone();
const json = responseClone.json();
return json
```
because then I get a 'Failed to execute 'clone' on 'Response': Response body is already used' error.

Any ideas how to go forward with this?

Thanks for any kind of help!

**Comment by simonxabris** [comment]
Hey!

I've implemented my solution described in the last comment's update block, opened a PR (https://github.com/andyrichardson/tipple/pull/25) with it, let me know what you think.

**Comment by andyrichardson** [comment]
Fixed in #25
